tests/integration/integration_nodejs_test.go

```diff
@@ -1310,3 +1311,23 @@ func TestCustomResourceTypeNameDynamicNode(t *testing.T) {
 		},
 	})
 }
+
+// Regression test for https://github.com/pulumi/pulumi/issues/12301
+func TestRegression12301Node(t *testing.T) {
+	integration.ProgramTest(t, &integration.ProgramTestOptions{
+		Dir:          filepath.Join("nodejs", "regression-12301"),
+		Dependencies: []string{"@pulumi/pulumi"},
+		PostPrepareProject: func(project *engine.Projinfo) error {
+			// Move the bad JSON file up one directory
+			jsonPath := filepath.Join(project.Root, "regression-12301.json")
+			dirName := filepath.Base(project.Root)
+			newPath := filepath.Join(project.Root, "..", dirName+".json")
+			return os.Rename(jsonPath, newPath)
+		},
+		ExtraRuntimeValidation: func(t *testing.T, stack integration.RuntimeValidationStackInfo) {
+			assert.Len(t, stack.Outputs, 1)
+			assert.Contains(t, stack.Outputs, "bar")
+			assert.Equal(t, 3.0, stack.Outputs["bar"].(float64))
```

Comment:
optional: we can collapse this into a single equals check:

```suggestion
			assert.Equal(t, map[string]any{"bar": float64(3)}, stack.Outputs)
```